### PR TITLE
Assert monotonic dump times #2103

### PR DESCRIPTION
**Pull request by ncm**
This is a pull request on behalf of Richard Myers, to ensure 
that dump is called with monotonically increasing timestamps.
It builds, but I have not run tests.

(BTW, in C++11 and later, line 51 in the header file could be

  vluint64_t m_lastDumpTime = 0;

and then not need to be initialized in the constructor.
Used faithfully, this can eliminate a lot of risk of 
uninitialized members.)

### CODE REVIEW

**Comment by RichardMyersGV** [comment]
I will test it...  Ideally, I think it should just return if timeui <= m_lastDumpTime, or return and put out a message.  My application is a DLL, and I don't see anything that would be output to the console, so I'd at least want it to do the right thing without wasting time re-writing FST traces (at best) or screwing up the trace file.

**Comment by ncm** [comment]
So, it looks like existing tests provoke the buggy multiple-dump behavior..

File: include/verilated_fst_c.cpp [hunk 1]
```diff
@@ -196,6 +198,9 @@ void VerilatedFst::addCallback(VerilatedFstCallback_t initcb, VerilatedFstCallba
 
 void VerilatedFst::dump(vluint64_t timeui) {
     if (!isOpen()) return;
+    assert(m_minNextDumpTime <= timeui);
+    if (   m_minNextDumpTime >  timeui) return;
```
Comment (wsnyder) [inline include/verilated_fst_c.cpp:202]: Nit: fix spacing

    if (m_minNextDumpTime > timeui) return;

Actually think this goes away with above assert.

File: include/verilated_fst_c.cpp [hunk 2]
```diff
@@ -35,6 +35,7 @@
 #include "gtkwave/lz4.c"
 
 #include <algorithm>
+#include <cassert>
```
Comment (wsnyder) [inline include/verilated_fst_c.cpp:38]: Remove

Comment (ncm) [reply include/verilated_fst_c.cpp]: Done

**Comment by wsnyder** [comment]
Please fix the Travis regression errors:

1. #dist/t_dist_contributors: %Error: Certify your contribution by sorted-inserting 'Nathan Myers' into docs/CONTRIBUTORS.

2. /home/travis/build/verilator/verilator/test_regress/../include/verilated_fst_c.cpp:202:62: error: unknown conversion type character ‘,’ in format [-Werror=format=]

I think you're missing the "u" and don't need the casts.

        VL_PRINTF_MT("%%Warning: previous dump at t=%" VL_PRI64 "u, requesting t=%" VL_PRI64 "u\n",
        m_minNextDumpTime - 1, timeui);

**Comment by ncm** [comment]
Oops, sorry to be a burden.

**Comment by wsnyder** [comment]
No worries. Thanks for seeing this through, I appreciate for so few lines there was a lot of churn, hopefully you'll have something large to patch at some point ;)

I'll merge assuming regressions look good.

**Comment by ncm** [comment]
I have not done a "git rebase -i master" to squeeze the diffs into one commit, because I don't know how github would treat it, or what you prefer. Assuming github handles it sensibly, do you have a policy or preference?

Whether you would like me to clean it up and push again, or would rather do it yourself, or not, I am ever your humble servant.

**Comment by wsnyder** [comment]
As long as github says no conflicts there's no need to rebase, and I'll always squash.

**Pull request merged**